### ISSUE TITLE
Command registration fails

### ISSUE THREAD

**Issue by leNicDev** [Alan-Gomes/mcspring-boot#3]
When the commands are registered, the following exception is thrown:
```
[21:40:32 INFO]: 2019-04-24 21:40:32.745 ERROR 13088 --- [  Server thread] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'commandInterceptor': Unsatisfied dependency expressed through field 'commandExecutor'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'defaultCommandExecutor': Unsatisfied dependency expressed through field 'cli'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'picocliCommandLine' defined in class path resource [dev/alangomes/springspigot/picocli/PicocliConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.alangomes.springspigot.picocli.CommandLineDefinition]: Factory method 'picocliCommandLine' threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:596) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:90) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:374) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1411) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:592) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:849) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at bz.dcr.deinbusiness.DeinBusinessPlugin.onEnable(DeinBusinessPlugin.java:56) [deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:265) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:347) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:410) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.enablePlugin(CraftServer.java:436) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.enablePlugins(CraftServer.java:350) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at net.minecraft.server.v1_13_R2.MinecraftServer.l(MinecraftServer.java:580) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:542) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:420) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at net.minecraft.server.v1_13_R2.DedicatedServer.init(DedicatedServer.java:294) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:698) [spigot-1.13.2.jar:git-Spigot-f09662d-494eef4]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212-1-ojdkbuild]
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'defaultCommandExecutor': Unsatisfied dependency expressed through field 'cli'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'picocliCommandLine' defined in class path resource [dev/alangomes/springspigot/picocli/PicocliConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.alangomes.springspigot.picocli.CommandLineDefinition]: Factory method 'picocliCommandLine' threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:596) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:90) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:374) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1411) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:592) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:277) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1247) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1167) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:593) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	... 27 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'picocliCommandLine' defined in class path resource [dev/alangomes/springspigot/picocli/PicocliConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.alangomes.springspigot.picocli.CommandLineDefinition]: Factory method 'picocliCommandLine' threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:627) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:607) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:277) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1247) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1167) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:593) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	... 40 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.alangomes.springspigot.picocli.CommandLineDefinition]: Factory method 'picocliCommandLine' threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:622) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	... 53 common frames omitted
Caused by: java.lang.NullPointerException: null
	at dev.alangomes.springspigot.picocli.PicocliConfiguration.getCommandName(PicocliConfiguration.java:46) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at dev.alangomes.springspigot.picocli.PicocliConfiguration.registerCommands(PicocliConfiguration.java:116) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at dev.alangomes.springspigot.picocli.PicocliConfiguration.picocliCommandLine(PicocliConfiguration.java:31) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at dev.alangomes.springspigot.picocli.PicocliConfiguration$$EnhancerBySpringCGLIB$$3eb02182.CGLIB$picocliCommandLine$0(<generated>) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at dev.alangomes.springspigot.picocli.PicocliConfiguration$$EnhancerBySpringCGLIB$$3eb02182$$FastClassBySpringCGLIB$$10b1aa8f.invoke(<generated>) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:363) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at dev.alangomes.springspigot.picocli.PicocliConfiguration$$EnhancerBySpringCGLIB$$3eb02182.picocliCommandLine(<generated>) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_212-1-ojdkbuild]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_212-1-ojdkbuild]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_212-1-ojdkbuild]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_212-1-ojdkbuild]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[deinbusiness-3.0.0-shaded.jar:3.0.0]
	... 54 common frames omitted
```

Command class:
```
package bz.dcr.deinbusiness.command;

import bz.dcr.deinbusiness.DeinBusinessPlugin;
import bz.dcr.deinbusiness.business.BusinessService;
import bz.dcr.deinbusiness.company.Company;
import bz.dcr.deinbusiness.company.CompanyMember;
import bz.dcr.deinbusiness.company.CompanyPosition;
import bz.dcr.deinbusiness.lang.LangKey;
import dev.alangomes.springspigot.context.Context;
import dev.alangomes.springspigot.security.Authorize;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import picocli.CommandLine;

import java.util.HashMap;
import java.util.Map;

@Component
@CommandLine.Command(name = "bemployees")
public class BusinessEmployeesCommand implements Runnable {

    private DeinBusinessPlugin plugin;
    private Context context;
    private BusinessService businessService;


    @Autowired
    public BusinessEmployeesCommand(Plugin plugin, Context context, BusinessService businessService) {
        this.plugin = (DeinBusinessPlugin) plugin;
        this.context = context;
        this.businessService = businessService;
    }


    @Authorize("hasPermission('deinbusiness.listemployees')")
    @Override
    public void run() {
        final Player player = context.getPlayer();

        Bukkit.getScheduler().runTaskAsynchronously(plugin, () -> {
            // Get players company
            final Company company = businessService.getCompany(player.getUniqueId());

            // Company does already exist
            if (company == null) {
                player.sendMessage(plugin.getLangManager().getMessage(LangKey.ERR_NO_COMPANY, true));
                return;
            }

            // Fill placeholders
            final Map<String, String> placeholders = new HashMap<>();
            placeholders.put("company", company.getName());
            placeholders.put("employee_amount", plugin.getLangManager().formatNumber(company.getMembers().size()));

            // Send header
            player.sendMessage(plugin.getLangManager()
                    .getMessage(LangKey.STATUS_EMPLOYEE_LIST_HEADER, false, placeholders));

            // Print employees
            int index = 1;
            CompanyPosition position;
            for (CompanyMember member : company.getMembers().values()) {
                position = company.getPosition(member.getPosition());

                // Member has invalid position
                if (position == null) {
                    plugin.getLogger().warning("Player " + member.getName() +
                            " has invalid position (" + member.getPosition().name() + ")!");
                    return;
                }

                // Fill placeholders
                placeholders.put("index", plugin.getLangManager().formatNumber(index));
                placeholders.put("employee", member.getName());
                placeholders.put("position", member.getPosition().name());
                placeholders.put("salary", plugin.getLangManager().formatNumber(position.getSalary()));

                // Send message
                player.sendMessage(plugin.getLangManager()
                        .getMessage(LangKey.STATUS_EMPLOYEE_LIST_ITEM, false, placeholders));

                index++;
            }
        });
    }

}
```

This happens with every command class.

**Comment by Alan-Gomes** [comment]
Hi, thanks for the feedback. The project has another class annotated with @CommandLine.Command? The error seems to be caused by a command without name.

**Comment by leNicDev** [comment]
There's definitely no command class without a name. I just added you as a collaborator to the repo:
https://github.com/leNicDev/deinBusiness3

**Comment by Alan-Gomes** [comment]
The issue was caused by the spring gclib, which gets activated when the `@Authorized` annotation is used. I released a patch on the version 0.7.0

Thanks for reporting this issue :smiley:

**Comment by leNicDev** [comment]
@Alan-Gomes Thanks for the fix! :) But there still seems to be an issue with the `@Authorized` annotation.
I can't get parameter injection working on a command annotated with `@Authorized`.

```
@CommandLine.Parameters(index = "0", description = "Random parameter")
    private String param;
```

In this case, `param` will always be null when the `run()` method has an `@Authorize` annotation.

**Comment by Alan-Gomes** [comment]
This is actually a limitation in picocli reflection system, so there's nothing much to do. I recommend putting the `@Authorize` annotations in another bean, such as a service

Since the annotation throws a `PermissionDeniedException`, it will works like a charm and it will also make the permission decoupled from the command classes :smile:

I will add a warning in the documentation about this limitation

Thanks

**Comment by Alan-Gomes** [comment]
> This is actually a limitation in picocli reflection system, so there's nothing much to do. I recommend putting the `@Authorize` annotations in another bean, such as a service
> 
> Since the annotation throws a `PermissionDeniedException`, it will works like a charm and it will also make the permission decoupled from the command classes smile
> 
> I will add a warning in the documentation about this limitation
> 
> Thanks

For reference, this issue has been fixed on https://github.com/Alan-Gomes/mcspring-boot/commit/fbb689cce6a3059ceb3ea9aee175b5950a583041 which will be released on 0.19.0

GCLib proxies are now supported on command classes.
